### ISSUE TITLE
Reader.ReadList throws ArgumentOutOfRange

### ISSUE THREAD

**Issue by ptobien** [FirstGearGames/FishNet#100]
I believe the following line doesn't work if `collection` is empty even if the capacity is increased beforehand:

https://github.com/FirstGearGames/FishNet/blob/60127890656ea50ba69f21de92db1dccc8e26eb1/Assets/FishNet/Runtime/Serializing/Reader.cs#L1106

Changing the line to

```cs
collection.Add(Read<T>());
```

resolves the issue but, as is, doesn't override existing items in the collection. I'm not sure why the list isn't re-instantiated in general though. Personally, I just changed the last few lines to:

```cs
...
else {
    collection = new List<T>(count);
    
    for (int i = 0; i < count; i++)
        collection.Add(Read<T>());
    
    return count;
}
```

**Comment by FirstGearGames** [comment]
You're right, it does need to fill the missing entries. It's not re-instantiated because the user specifies the list and they may not want it wiped. I'll have this fixed in the next release.

**Comment by ptobien** [comment]
Alright, I see. Thanks a lot! Great project by the way.

**Comment by FirstGearGames** [comment]
I took this information into consideration and I believe most would expect the list to clear. I cannot think of a situation where keeping old values would make sense. I'll update to clear and add entries.

**Comment by FirstGearGames** [comment]
How do we feel about this?
`        public int ReadList<T>(ref List<T> collection)
        {
            int count = ReadInt32();
            if (count == -1)
            {
                return 0;
            }
            else
            {
                if (collection == null)
                    collection = new List<T>(count);
                collection.Clear();

                for (int i = 0; i < count; i++)
                    collection.Add(Read<T>());

                return count;
            }
        }`

**Comment by ptobien** [comment]
Looks good to me.

Maybe one small remark as I had a quick look at `Writer.WriteList`. I believe WriteList writes `-1` as "count" if the passed in list is null (and only if it is null). Wouldn't it make sense if the Reader also returns null for the list in that case? So:

```cs
public int ReadList<T>(ref List<T> collection)
{
    int count = ReadInt32();
    if (count == -1)
    {
        // the collection was null during serialization
        // so we make sure that it is deserialized to null as well
        collection = null;
        return 0;
    }
    else
    {
        if (collection == null)
            collection = new List<T>(count);
        else
            collection.Clear();

        for (int i = 0; i < count; i++)
            collection.Add(Read<T>());

        return count;
    }
}
```

I think I would generally expect that serializing "null" would also deserialize to null so this seems more consistent to me. I might be overlooking something here though. What do you think?

**Comment by FirstGearGames** [comment]
I did consider that when creating this method but setting it to null would cause the user to have to reinitialize it, resulting in GC.

**Comment by ptobien** [comment]
Yeah, that's true. Unless you want it to be null for whatever reason (which I'm kinda assuming here since serializing a collection that's null seems like a thing you would do on purpose).

I'm actually using `ReadListAllocated` which has the expected behaviour either way (I'm typically not interested in the count).

**Comment by ptobien** [comment]
Sure, sounds good to me! Thanks for the quick update! Looking forward to the next release.

**Comment by FirstGearGames** [comment]
Put in as this. Will be available in 2.4.1.

```
        /// <summary>
        /// Reads into collection and returns amount read.
        /// </summary>
        /// <param name="collection"></param>
        /// <param name="allowNullification">True to allow the referenced collection to be nullified when receiving a null collection read.</param>
        /// <returns>Number of values read into the collection. -1 is returned if the collection were read as null.</returns>
        [CodegenExclude]
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public int ReadList<T>(ref List<T> collection, bool allowNullification = false)
        {
            int count = ReadInt32();
            if (count == -1)
            {
                if (allowNullification)
                    collection = null;
                return -1;
            }
            else
            {
                if (collection == null)
                    collection = new List<T>(count);
                else
                    collection.Clear();

                for (int i = 0; i < count; i++)
                    collection.Add(Read<T>());

                return count;
            }
        }
        ```
